File: calco/state.py
from dataclasses import dataclass
from pyrsistent import PSet

from .defs import Attr, Prop


@dataclass
class State:
    attrs: PSet
    props: PSet

    def __init__(self, attrs=None, props=None):
        self.attrs = {} if attrs is None else attrs
        self.props = {} if props is None else props

    def union(self, s: State):
        return State(self.attrs.union(s.attrs), self.props.union(s.props))

File: src/Calco/State.hs
{-# LANGUAGE ConstraintKinds    #-}
{-# LANGUAGE GADTs              #-}
{-# LANGUAGE StandaloneDeriving #-}

module Calco.State where

import           Data.Set   (Set)
import qualified Data.Set   as Set

import           Calco.Defs

type StateContext a p = (Attr a, Prop p)

-- State of the stream on the graph's edge.
data State a p where
  State :: StateContext a p
        => { attrs :: Set a
           , props :: Set p }
        -> State a p

deriving instance Show (State a p)

instance StateContext a p => Semigroup (State a p) where
  (<>) = union

instance StateContext a p => Monoid (State a p) where
  mempty = empty

empty :: StateContext a p => State a p
empty = State { attrs = Set.empty
              , props = Set.empty }

union :: StateContext a p => State a p -> State a p -> State a p
union s1 s2 = State { attrs = attrs s1 <> attrs s2
                    , props = props s1 <> props s2 }
